File: Google-Photos-API/PhotosSample/app/src/main/java/com/example/nabha/photossample/PhotosAdapter.java
package com.example.nabha.photossample;

import android.graphics.Bitmap;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.google.gdata.data.photos.PhotoEntry;
import com.squareup.picasso.Picasso;

import java.net.URL;
import java.util.List;

public class PhotosAdapter extends RecyclerView.Adapter<PhotosAdapter.PhotoViewHolder> {

    private static final String TAG = Utils.TAG;
    private List<PhotoEntry> mPhotoEntries;
    private Picasso mPicasso;

    PhotosAdapter(List<PhotoEntry> photoEntries, Picasso picasso) {
        mPhotoEntries = photoEntries;
        mPicasso = picasso;
    }


    @NonNull
    @Override
    public PhotoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.single_item_layout, parent, false);
        return new PhotoViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull PhotoViewHolder holder, int position) {
        mPicasso.load(mPhotoEntries.get(position).getMediaContents().get(0).getUrl())
                .placeholder(android.R.drawable.gallery_thumb)
                .error(android.R.drawable.gallery_thumb)
                .into(holder.imageView);
    }

    @Override
    public int getItemCount() {
        return mPhotoEntries.size();
    }

    /*private Bitmap getBitmap(String urlString) {
        Bitmap bmp = null;
        try {
            URL photoUrl = new URL(urlString);
            bmp = BitmapFactory.decodeStream(photoUrl.openConnection().getInputStream());
        } catch (Exception e) {
            e.printStackTrace();
        }
        return bmp;
    }*/

    private void getBitmap(final PhotoViewHolder holder, String urlString) {
        try {
            URL photoUrl = new URL(urlString);
            new GenerateBitmap(holder.getAdapterPosition(), photoUrl, new GenerateBitmap.Callback() {
                @Override
                public void onBitmapReceived(int position, Bitmap bitmap) {
                    if (holder.getAdapterPosition() == position) {
                        holder.imageView.setImageBitmap(bitmap);
                    }
                }
            }).execute();
        } catch (Exception ignored) {}
    }

    class PhotoViewHolder extends RecyclerView.ViewHolder {

        ImageView imageView;

        public PhotoViewHolder(View itemView) {
            super(itemView);
            imageView = itemView.findViewById(R.id.photo_image);
        }
    }
}

File: KProject8/app/src/main/java/com/example/nabha/k_project_8/adapter/PhotosAdapter.kt
package com.example.nabha.k_project_8.adapter

import android.graphics.Bitmap
import android.graphics.Color
import android.os.AsyncTask
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import com.example.nabha.k_project_8.R
import com.example.nabha.k_project_8.model.Photo
import com.example.nabha.k_project_8.utils.ThumbnailGenerator

class PhotosAdapter(val photosList: ArrayList<Photo>, val listener: IClickListener) :
        RecyclerView.Adapter<PhotosAdapter.PhotoViewHolder>(), ThumbnailGenerator.IImageReceiverCallback{
    private val thumbnailGenerator = ThumbnailGenerator(this)
    private var bitmap : Bitmap? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PhotoViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.single_item_photo, parent, false)
        return PhotoViewHolder(view)
    }

    override fun getItemCount(): Int {
        return photosList.size
    }

    override fun onBindViewHolder(holder: PhotoViewHolder, position: Int) {
        thumbnailGenerator.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, photosList[position].path, position.toString())
        if (bitmap != null) {
            holder.thumbnail?.setImageBitmap(bitmap)
        } else {
            holder.thumbnail?.setColorFilter(Color.BLACK)
        }
        holder.thumbnail?.setOnClickListener {
            listener.onItemClick(photosList[position].path, position)
        }
    }

    override fun onThumbnailReceived(thumbnail: Bitmap, path: String, position: Int) {
        bitmap = thumbnail
        notifyItemChanged(position)
    }

    class PhotoViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val thumbnail: ImageView? = itemView.findViewById(R.id.thumbnailImage)
    }

    interface IClickListener {
        fun onItemClick(path: String, position: Int)
    }
}